core/spring/src/main/java/org/apache/syncope/core/spring/ImplementationManager.java

```diff
@@ -104,10 +104,19 @@
                         rule = (AccountRule) ApplicationContextProvider.getBeanFactory().
                                 getSingleton(ruleClass.getName());
                     } else {
-                        rule = (AccountRule) ApplicationContextProvider.getBeanFactory().
-                                createBean(ruleClass, AbstractBeanDefinition.AUTOWIRE_BY_TYPE, false);
-                        ApplicationContextProvider.getBeanFactory().
-                                registerSingleton(ruleClass.getName(), rule);
+                        // to avoid concurrency issues
+                        synchronized (ImplementationManager.class) {
+                            // double check for performance reasons
+                            if (ApplicationContextProvider.getBeanFactory().containsSingleton(ruleClass.getName())) {
```

Comment:
Why to repeat this check, which was already performed just a few lines above?
You can refactor the whole block as follows, simplify the logic and avoid duplicates:

```java
                // fetch (or create) rule
                synchronized (LOG) {
                    if (ApplicationContextProvider.getBeanFactory().containsSingleton(ruleClass.getName())) {
                        rule = (AccountRule) ApplicationContextProvider.getBeanFactory().
                                getSingleton(ruleClass.getName());
                    } else {
                                rule = (AccountRule) ApplicationContextProvider.getBeanFactory().
                                        createBean(ruleClass, AbstractBeanDefinition.AUTOWIRE_BY_TYPE, false);
                                ApplicationContextProvider.getBeanFactory().
                                        registerSingleton(ruleClass.getName(), rule);
                        }
                }
```

Reply:
assumption: 
* synchronized is expensive and therefore I want to use it as less as possible
* I want to register the bean exactly one time

If you do it like this, every time a rule is loaded its synchronized. This means it's in a sequential mode. and we do it exclusively every time. This works but is also a lack of performance

If you do it with two checks,  you'll get the first check as usual. the will only be executed from the first few callers. so, only they pay the price for synchronized.

Some references (this is more for a singleton but same patterns apply) #5: https://www.geeksforgeeks.org/java-singleton-design-pattern-practices-examples/
also here https://stackoverflow.com/questions/11165852/java-singleton-and-synchronization

Reply:
So the logic is:

1. check if there is a singleton bean
    1. if so, take it
    1. if not, synchronize and check again
          1. if so, take it
          1. if not, create it

Also, did you take into account that [DefaultSingletonBeanRegistry#containsSingleton](https://github.com/spring-projects/spring-framework/blob/master/spring-beans/src/main/java/org/springframework/beans/factory/support/DefaultSingletonBeanRegistry.java#L285) is backed by a `ConcurrentHashMap`?

About the provided references, I am not convinced they apply to the case as above.